app/src/main/java/org/thoughtcrime/securesms/components/TransferControlView.java

```diff
@@ -0,0 +1,257 @@
+
+    display(progressWheel);
+  }
+
+  public void setDownloadClickListener(final @Nullable OnClickListener listener) {
+    downloadDetails.setOnClickListener(listener);
+  }
+
+  public void clear() {
+    clearAnimation();
+    setVisibility(GONE);
+    if (current != null) {
+      current.clearAnimation();
+      current.setVisibility(GONE);
+    }
+    current = null;
+    slides = null;
+  }
+
+  public void setShowDownloadText(boolean showDownloadText) {
+    downloadDetailsText.setVisibility(showDownloadText ? VISIBLE : GONE);
+    forceLayout();
+  }
+
+  private boolean isUpdateToExistingSet(@NonNull List<Slide> slides) {
+    if (slides.size() != networkProgress.size()) {
+      return false;
+    }
+
+    for (Slide slide : slides) {
+      if (!networkProgress.containsKey(slide.asAttachment())) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  private int getTransferState(@NonNull List<Slide> slides) {
+    int transferState = AttachmentDatabase.TRANSFER_PROGRESS_DONE;
+    for (Slide slide : slides) {
+      if (slide.getTransferState() == AttachmentDatabase.TRANSFER_PROGRESS_PENDING && transferState == AttachmentDatabase.TRANSFER_PROGRESS_DONE) {
+        transferState = slide.getTransferState();
+      } else {
+        transferState = Math.max(transferState, slide.getTransferState());
+      }
+    }
+    return transferState;
+  }
+
+  private String getDownloadText(@NonNull List<Slide> slides) {
+    if (slides.size() == 1) {
+      return slides.get(0).getContentDescription();
+    } else {
+      int downloadCount = Stream.of(slides).reduce(0, (count, slide) -> slide.getTransferState() != AttachmentDatabase.TRANSFER_PROGRESS_DONE ? count + 1 : count);
+      return getContext().getResources().getQuantityString(R.plurals.TransferControlView_n_items, downloadCount, downloadCount);
+    }
+  }
+
+  private void display(@Nullable final View view) {
+    if (current == view) {
+      return;
+    }
+
+    if (current != null) {
+      current.setVisibility(GONE);
+    }
+
+    if (view != null) {
+      view.setVisibility(VISIBLE);
+      setVisibility(VISIBLE);
+    } else {
+      setVisibility(GONE);
+    }
+
+    current = view;
+  }
+
+  private static float calculateProgress(@NonNull Map<Attachment, Float> uploadDownloadProgress, Map<Attachment, Float> compresssionProgress) {
+    float totalDownloadProgress    = 0;
+    float totalCompressionProgress = 0;
+
+    for (float progress : uploadDownloadProgress.values()) {
+      totalDownloadProgress += progress;
+    }
+
+    for (float progress : compresssionProgress.values()) {
+      totalCompressionProgress += progress;
+    }
+
+    float weightedProgress = UPLOAD_TASK_WEIGHT * totalDownloadProgress         + COMPRESSION_TASK_WEIGHT * totalCompressionProgress;
+    float weightedTotal    = UPLOAD_TASK_WEIGHT * uploadDownloadProgress.size() + COMPRESSION_TASK_WEIGHT * compresssionProgress.size();
+
+    return weightedProgress / weightedTotal;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).